x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/transport/TransportFindFieldStructureAction.java

```diff
@@ -0,0 +1,214 @@
+            builder.startObject();
+            for (ResourcePrivileges index : response.getIndexPrivileges()) {
+                builder.field(index.getResource());
+                builder.map(index.getPrivileges());
+            }
+            builder.endObject();
+
+            listener.onFailure(
+                Exceptions.authorizationError(
+                    "Cannot determine field structure because user {} lacks permissions on the indices: {}",
+                    username,
+                    Strings.toString(builder)
+                )
+            );
+        }
+    }
+
+    private void searchAndBuildTextStructureResponse(
+        FindFieldStructureAction.Request request,
+        Map<String, String> headers,
+        ActionListener<TextStructureResponse> listener
+    ) {
+
+        ActionListener<SearchResponse> searchResponseActionListener = ActionListener.wrap(searchResponse -> {
+            // TODO could we just feed the doc fields as "lines" to the field structure finder?
+            StringBuilder stringBuilder = new StringBuilder();
+            for (SearchHit hit : searchResponse.getHits().getHits()) {
+                if (request.getRuntimeMappings().isEmpty()) {
+                    Object value = MapHelper.dig(request.getFieldName(), Objects.requireNonNull(hit.getSourceAsMap()));
+                    if (value instanceof String) {
+                        stringBuilder.append(value.toString()).append("\n");
+                    } else if (value instanceof List<?>) {
+                        @SuppressWarnings("unchecked")
+                        List<Object> values = (List<Object>) value;
+                        for (Object v : values) {
+                            if (v instanceof String) {
+                                stringBuilder.append(v.toString()).append("\n");
+                            }
+                        }
+                    }
+                } else {
+                    DocumentField field = hit.field(request.getFieldName());
+                    for (Object v : field.getValues()) {
+                        if (v instanceof String) {
+                            stringBuilder.append(v.toString()).append("\n");
+                        }
+                    }
+                }
+            }
+            threadPool.executor(GENERIC).execute(() -> {
+                try {
+                    listener.onResponse(buildTextStructureResponse(request, new BytesArray(stringBuilder.toString())));
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                }
+            });
+        }, listener::onFailure);
+        FunctionScoreQueryBuilder functionBuilder = request.getQueryBuilder() != null
+            ? QueryBuilders.functionScoreQuery(request.getQueryBuilder(), ScoreFunctionBuilders.randomFunction())
+            : QueryBuilders.functionScoreQuery(ScoreFunctionBuilders.randomFunction());
```

Comment:
I think getting a random subsample based on the query is a good thing. It broadens our sampling to make the best possible guess given ALL the data.